parsers/verilog.c

```diff
@@ -1224,49 +1230,79 @@ static void processTypedef (tokenInfo *const token)
 	createTag (token);
 }
 
-static void processClass (tokenInfo *const token)
+static tokenInfo * processParameterList (int c)
 {
-	/*Note: At the moment, only identifies typedef name and not its contents */
-	int c;
-	tokenInfo *extra;
+	tokenInfo *token = newToken ();
+	tokenInfo *head = NULL;
 	tokenInfo *parameters = NULL;
-
-	/* Get identifiers */
-	c = skipWhite (vGetc ());
-	if (isIdentifierCharacter (c))
-	{
-		readIdentifier (token, c);
-		c = skipWhite (vGetc ());
-	}
-
-	/* Find class parameters list */
 	if (c == '#')
 	{
 		c = skipWhite (vGetc ());
 		if (c == '(')
 		{
-			parameters = newToken ();
 			do
 			{
 				c = skipWhite (vGetc ());
-				readIdentifier (parameters, c);
-				updateKind (parameters);
-				verbose ("Found class parameter %s\n", vStringValue (parameters->name));
-				if (parameters->kind == K_UNDEFINED)
+				if (isIdentifierCharacter (c))
 				{
-					parameters->kind = K_CONSTANT;
-					parameters = pushToken (parameters, newToken ());
-					c = vGetc();
-					while (c != ',' && c != ')' && c != EOF)
+					readIdentifier (token, c);
+					updateKind (token);
+					verbose ("Found parameter %s\n", vStringValue (token->name));
+					if (token->kind == K_UNDEFINED)
 					{
-						c = vGetc();
+						c = skipWhite (vGetc ());
+						if (c == ',' || c == ')' || c == '=')	// ignore user defined type
+						{
+							token->kind = K_CONSTANT;
+							if (head == NULL)
+							{
+								head = token;
+								parameters = token;
+							} else
+								parameters = appendToken (parameters, token);	// append token on parameters
+							token = newToken ();
+							// skip expression
+							while (c != ',' && c != ')' && c != EOF)
+							{
+								if (c == '(')
+									c = skipPastMatch ("()");
+								else
+									c = skipWhite (vGetc ());
+							}
+						}
 					}
 				}
+				else if (c == '[') {
+					c =skipDimension(c);
+					vUngetc (c);
+				}
 			} while (c != ')' && c != EOF);
 			c = skipWhite (vGetc ());
-			parameters = popToken (parameters);
+			deleteToken (token);
 		}
 	}
+	vUngetc (c);
```

Comment:
if c is not '#', the memory object associated with token leaks.

1. install valgrind
2. git checkout a54e815ca8dbfcd96e0eae972191e74e6aa684a2
3. make
4. make units LANGUAGES=SystemVerilog VG=1
5. cat Units/parser-verilog.r/systemverilog-class.d/VALGRIND.tmp 
```
==20931== Memcheck, a memory error detector
==20931== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==20931== Using Valgrind-3.14.0 and LibVEX; rerun with -h for copyright info
==20931== Command: ./ctags --verbose --options=NONE --optlib-dir=+./Units/parser-verilog.r/systemverilog-class.d/optlib -o - --options=./Units/parser-verilog.r/systemverilog-class.d/args.ctags ./Units/parser-verilog.r/systemverilog-class.d/input.sv
==20931== Parent PID: 19254
==20931== 
==20931== 
==20931== HEAP SUMMARY:
==20931==     in use at exit: 768 bytes in 21 blocks
==20931==   total heap usage: 6,777 allocs, 6,756 frees, 360,392 bytes allocated
==20931== 
==20931== 768 (264 direct, 504 indirect) bytes in 3 blocks are definitely lost in loss record 7 of 7
==20931==    at 0x4C2EE3B: malloc (vg_replace_malloc.c:309)
==20931==    by 0x4149BD: eMalloc (routines.c:220)
==20931==    by 0x471C8E: newToken (verilog.c:380)
==20931==    by 0x472147: processParameterList (verilog.c:1235)
==20931==    by 0x4736A8: processClass (verilog.c:1304)
==20931==    by 0x4736A8: findTag (verilog.c:1577)
==20931==    by 0x4736A8: findVerilogTags (verilog.c:1648)
==20931==    by 0x411292: createTagsForFile (parse.c:3722)
==20931==    by 0x411292: createTagsWithFallback1 (parse.c:3844)
==20931==    by 0x4115E3: createTagsWithFallback (parse.c:3934)
==20931==    by 0x4115E3: parseMio (parse.c:4096)
==20931==    by 0x411746: parseFileWithMio (parse.c:4142)
==20931==    by 0x411873: parseFile (parse.c:4079)
==20931==    by 0x403307: createTagsForEntry (main.c:221)
==20931==    by 0x40364F: createTagsForArgs (main.c:266)
==20931==    by 0x40364F: batchMakeTags (main.c:360)
==20931==    by 0x403AD5: runMainLoop (main.c:332)
==20931==    by 0x403AD5: ctags_cli_main (main.c:585)
==20931== 
==20931== LEAK SUMMARY:
==20931==    definitely lost: 264 bytes in 3 blocks
==20931==    indirectly lost: 504 bytes in 18 blocks
==20931==      possibly lost: 0 bytes in 0 blocks
==20931==    still reachable: 0 bytes in 0 blocks
==20931==         suppressed: 0 bytes in 0 blocks
==20931== 
==20931== For counts of detected and suppressed errors, rerun with: -v
==20931== ERROR SUMMARY: 1 errors from 1 contexts (suppressed: 0 from 0)
```

Reply:
The GitHub action "run units target under VALGRIND" looks broken. It should report this memory leak. I must fix it.